### ISSUE TITLE
Class Components do not populate their ref attribute

### ISSUE THREAD

**Issue by haferje** [alex-kinokon/jsx-dom#104]
I am trying to use `createRef()` to get a reference to a class component instance, so that I can call methods on it, but the `current` value always returns `null`.

This is an over-simplified version of the code, but gives an idea as to the issue I am running into.

```ts
class Page {
	protected button = createRef();

	constructor(root: HTMLElement) {
		root.append(<ClassComponentButton ref={this.button} />);
	}

	clickButton(): void {
		this.button.current.click();
	}
}
```

When the `clickButton()` method of `Page` is called, `this.button.current` is always `null`.

I have some experience with React in the past, but not familiar with many of the nuances, so I am sorry if I'm expecting certain functionality, incorrectly.  Reading React documentation, I should be able to use `createRef()` and `ref=` attributes on class components to get a reference to their instance.  I have even tried to use the callback version of the ref attribute, but the callback is never triggered.

Thank you.

This is the only way I can seem to get this working:

```ts
class Page {
	protected button;

	constructor(root: HTMLElement) {
		this.button = new ClassComponentButton();
		root.append(this.button.render());
	}

	clickButton(): void {
		this.button.click();
	}
}
```

Am I witnessing a bug, or am I going about this the wrong way? Should I be looking at `forwardRef` instead?  I understand jsx-dom isn't a one-to-one replacement for React, so I might just be looking for some guidance on doing this correctly.

This has been a great library for our use so far.
Thank you!

EDIT:
here is a more comprehensive, working example https://codesandbox.io/p/sandbox/jsx-dom-test-rcky7t?file=%2Fsrc%2FApp.tsx%3A45%2C10

**Comment by alex-kinokon** [comment]
I don’t have permission to access your sandbox link.

**Comment by volkanceylan** [comment]
I am not sure but I think the problem might be because you are not using render method but creating it in the constructor. AFAIK ref callback is executed during render phase for class components.

**Comment by haferje** [comment]
So do ref attributes on React class components not work, when attached in standard class implementations? They seem to work fine on standard DOM tags in the same scenario...

Sorry for the over simplified example. I updated my example link so it should be publicly available now.

**Comment by volkanceylan** [comment]
As I said, I am not sure, when I look at jsx dom code there is only two places attachRef is called, one when creating html elements and one under something related to shadow dom. In normal react/preact process refs are attached after render, in the mount phase. As jsx dom has no such phase maybe it should do it in initClassComponent method. Anyway, it should do it after calling render, the author should know better if I am missing something. You are rendering sub tree in constructor which I think is wrong in React class component system, but ref issue is probably not related to that

**Comment by alex-kinokon** [comment]
published as 8.1.4

**Comment by haferje** [comment]
That fixed it!
What a quick turn around - thank you both!
